\name{barplot-lattice}
\alias{barplot-lattice}
\alias{barplot.lattice.counts}
\alias{barplot.lattice.fraction}


\title{lattice barplot inner functions for ASEset objects}
\description{
	Generates lattice barplots for ASEset objects. Two levels of plotting detail are provided: a detailed barplot of read counts by allele useful for fewer samples and SNPs, and a less detailed barplot of the fraction of imbalance, useful for more samples and SNPs. 
}
\usage{
  barplot.lattice.fraction(
	identifier, 
	afraction,
	arank,
	amainVec, 
	... 
	)

  barplot.lattice.counts(
	identifier,
	acounts,
	arank,
	amainVec,
	...
	)
}

\arguments{
	\item{identifier,}{the single snp name to plot}
	\item{afraction}{the fractions in matrix for each snp}
	\item{arank}{the ranks for each snp}
	\item{acounts}{the counts for each snp}
	\item{amainVec,  }{vector for each plots main}
	\item{...}{for simpler generics when extending function}
}	


\details{
	\code{filter.pValue.fraction} is intended to remove p-value annotation with very large difference in frequency, which could just be a sequencing mistake. This is to avoid p-values like 1e-235 or similar.

	\code{sampleColour}{User specified colours, either given as named colours ('red', 'blue', etc) or as hexadecimal code. Can be either length 1 for all samples, or else of a length corresponding to the number of samples for individual colouring.}
}

\author{Jesper R. Gadin, Lasse Folkersen}

\seealso{
	\itemize{
		\item The \code{\link{ASEset}} class which the barplot function can be called up on.
	}
}

\examples{
	a <- ASEset
	strand <- "+"
	acounts <-  alleleCounts(a,strand=strand)
	arank <-  arank(a,strand=strand)
	afraction <- fraction(a, strand=strand)
	mainVec <- rep("",nrow(x))

	name <- rownames(a)[1]

	barplot.lattice.fraction(identifier=name, afraction, arank, amainVec) 
	barplot.lattice.counts(identifier=name,  acounts, arank, amainVec) 

}
\keyword{barplot}

